kiwi/builder/disk.py

```diff
@@ -1322,14 +1322,15 @@ def _install_bootloader(
 
         if self.bootloader != 'custom':
             # create bootloader config prior bootloader installation
-            self.bootloader_config.setup_disk_image_config(
-                boot_options=custom_install_arguments
-            )
-            if 's390' in self.arch:
-                self.bootloader_config.write()
-
-            # cleanup bootloader config resources taken prior to next steps
-            del self.bootloader_config
+            try:
+                self.bootloader_config.setup_disk_image_config(
+                    boot_options=custom_install_arguments
+                )
+                if 's390' in self.arch:
+                    self.bootloader_config.write()
+            finally:
```

Comment:
Note that without this try/finally wrap of `bootloader_config` calls if KIWI is aborted during `bootloader_config.setup_disk_image_config` or `bootloader_config.write` the bootloader_config instance is not immediately cleaned up as it would happen in a successful execution.

Reply:
Thanks I understand and makes a lot sense.

I think the entire disk builder should be refactored. This is imho the most complicated builder we have and it's no longer a straight forward read as it was when it was implemented back in time. It has grown to a complexity level where I think a refactor would make sense also with regards to the resource cleanup. Maybe we can think of a kiwi10 branch from where we also don't need to stay backward compatible and introduce a new disk builder. Just an idea

Reply:
Thanks I understand and makes a lot sense.

I think the entire disk builder should be refactored. This is imho the most complicated builder we have and it's no longer a straight forward read as it was when it was implemented back in time. It has grown to a complexity level where I think a refactor would make sense also with regards to the resource cleanup. Maybe we can think of a kiwi10 branch from where we also don't need to stay backward compatible and introduce a new disk builder. Just an idea